src/main/java/com/fouristhenumber/utilitiesinexcess/common/blocks/BlockDrum.java

```diff
@@ -51,50 +55,205 @@ public boolean hasTileEntity(int metadata) {
     @Override
     public boolean onBlockActivated(World world, int x, int y, int z, EntityPlayer player, int side, float hitX,
         float hitY, float hitZ) {
-        if (world.isRemote) {
+
+        TileEntity te = world.getTileEntity(x, y, z);
+
+        // all the drum things in variables so i can just ref them later xd
+        if (!(te instanceof TileEntityDrum drum)) return false;
+        FluidTank drumTank = drum.tank;
+        FluidStack fluid = drumTank.getFluid();
+        int capacity = drumTank.getCapacity();
+
+        ItemStack itemStack = player.inventory.getCurrentItem();
+        if (itemStack == null) {
+            if (world.isRemote) {
+                return false;
+            }
+
+            // Print drum capacity to player chat
+            player.addChatComponentMessage(
+                new ChatComponentTranslation(
+                    "%s %s",
+                    NumberFormat.DEFAULT.format(drumTank.getFluidAmount()),
+                    DrumConfig.unitToDisplay ? "mB" : "L"));
+
             return true;
         }
-        TileEntity tile = world.getTileEntity(x, y, z);
-        if (tile instanceof TileEntityDrum drum) {
-            ItemStack heldItem = player.getCurrentEquippedItem();
-            FluidStack heldFluid = FluidContainerRegistry.getFluidForFilledItem(heldItem);
+        Item item = itemStack.getItem();
 
-            if (FluidContainerRegistry.isFilledContainer(heldItem)) {
+        if (fluid == null) {
+            // empty drum cases
+            if (item instanceof IFluidContainerItem fluidTank) {
+                FluidStack playerFluid = fluidTank.getFluid(itemStack);
+                int playerCapacity = fluidTank.getCapacity(itemStack);
 
-                if (drum.tank.getFluid() == null) {
-                    drum.setFluid(new FluidStack(heldFluid.getFluid(), 0));
+                if (playerFluid == null) {
+                    // both the tank and the drum are false
+                    return true;
+                } else {
+                    // either the capacity of the drum, or the amount in the player hand
+                    int fillAmount = Math.min(capacity, playerFluid.amount);
+                    // copy the fluid add hte amount and fill the drum and empty the hand
+                    FluidStack fillFluid = playerFluid.copy();
+                    fillFluid.amount = fillAmount;
+                    int filedAmount = drumTank.fill(fillFluid, true);
+                    fluidTank.drain(itemStack, filedAmount, true);
+
+                    // todo: pull this to its own helper method?
+                    if (!world.isRemote) {
+                        player.addChatComponentMessage(
+                            new ChatComponentTranslation(
+                                "message.drum.filled",
```

Comment:
Please call the translation keys something that's closer to the convention. I see that there are already ones in the lang file under "tile.drum"